### ISSUE TITLE
Frequently when working on a Flutter app the tool will end with "Application finished." when it hasn't

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#8876]
Frequently when working on a Flutter app the tool will print "Application finished." and quit, with no other message or error or anything, but the app will still be running on the device.

It's not clear why.

This hurts the developer experience significantly.

**Comment by johnmccutchan** [comment]
@Hixie do you have any tips on reproducing this? How frequently are you experiencing this?

**Comment by Hixie** [comment]
I see this several times a day. My method of reproducing it is just "try to develop framework changes while using hot reload". I'm happy to run under a more verbose mode if that would help. I'll start running with -v, see if that gives me any more.

**Comment by Hixie** [comment]
I just caught this with `-v`. It doesn't give anything useful, unfortunately. The log ended with:
```
[  +42 ms] Connected to _flutterView/0xa4dfb4f0.
[        ] 🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
[        ] The Observatory debugger and profiler is available at: http://127.0.0.1:8100/
[        ] For a more detailed help message, press "h". To quit, press "q".
[+22829 ms] Application finished.
[   +1 ms] DevFS: Deleted filesystem on the device (file:///data/user/0/io.flutter.examples.Layers/cache/layerspfsrer/layers/)
[ +252 ms] TimeoutException after 0:00:00.250000: Future not completed
[        ] 'flutter run' took 99187ms.
[  +32 ms] ensureAnalyticsSent: 31ms
[   +1 ms] exiting with code 0
```

**Comment by johnmccutchan** [comment]
@Hixie is the app is still running/functional on the device? Based on the log it would appear:

1. We lost our connection to the service protocol.
2. We decide the app has exited.
3. We try and clean up DevFS
4. DevFS cleanup times out because the service protocol went down.

If the application is really running we should instrument the service protocol server and see if we can figure out why we dropped the connection.

**Comment by Hixie** [comment]
Yes, the app on the device is fine. Is there some way I can instrument the protocol for you during my usual usage? (I can't reproduce this on demand, so I have to leave it running with the right flags and stuff to catch it in the act. Some days it's worse than others.)

**Comment by yjbanov** [comment]
Assigning this issue to @johnmccutchan so that nobody else attempts to work on it simultaneously.

BTW, it seems the swallowing of the TimeoutException [is done intentionally](https://github.com/flutter/flutter/blob/7b2367ed5fa6cc596c2ae04d09364c4deed9d227/packages/flutter_tools/lib/src/run_hot.dart#L289).

**Comment by yjbanov** [comment]
Also moved it to "In Progress" in [M2](https://github.com/flutter/flutter/projects/10).

**Comment by eseidelGoogle** [comment]
I'm not sure how to move this bug forward.  Tempted to punt out of M2?  @Hixie and @xster have reported this occasionally.

**Comment by eseidelGoogle** [comment]
Probably what is needed is more --verbose logging to debug?

**Comment by Hixie** [comment]
It may be that what we need to do is be more aggressive about treating `adb` (and the equivalent on iOS) as unreliable, trying to reopen connections, etc, where we can.

**Comment by yjbanov** [comment]
I think the problem is that we're simply using the wrong message. AFAICT there are three ways to reach the `appFinished` method:

1. from `stopApp`, which corresponds to hitting the `q` key in `flutter run`
1. from `detach`, which corresponds to hitting the `d` key in `flutter run`
1. from `connectToServiceProtocol` when the VM service connection is closed (whether on purpose or not)

I think the first step is to correct our messages. Of the three cases above only the first one actually "finishes" the app. In the other two cases the app continues running. If we find that the issue is due to unexpected VM service interruptions, we should file a bug to fix that.

**Comment by yjbanov** [comment]
@johnmccutchan I can give it a shot if you are not working on it and/or have other stuff to work on.

**Comment by johnmccutchan** [comment]
@yjbanov We already know from this bug report that it isn't case 1. or case 2. (Hixie isn't detaching or quitting) I have adjusted the verbose logging to try and gain more information in case 3.

**Comment by johnmccutchan** [comment]
Also, the VM service server code will print a message to the console in the case that server goes down. I haven't seen that message in any of the logs. Likely issues:

1. The vm service isolate died. @Hixie next time you encounter this can you look at the logs and see if you can connect to Observatory on the device? If you can we know that the vm service is running fine.

2. Something about the adb port forwarding has gone wrong. A simple check is to connect to the device via `adb shell` and try and locally connect to the device's vm service port. If we can, than the forwarding provided by adb is the cause.

**Comment by yjbanov** [comment]
From my experience in the devicelab `adb` isn't very reliable, which is why we have code [like this](https://github.com/flutter/cocoon/blob/338d27c2af4ec0edaa81e866849661ce34bd9cea/agent/lib/src/adb.dart#L130) in the lab.

**Comment by johnmccutchan** [comment]
@yjbanov Interesting. I wonder what happens to port forwarding across adb restarts?

**Comment by johnmccutchan** [comment]
@yjbanov if we can reactivate the port forwarding, we should be able reconnect to the service protocol.

**Comment by yjbanov** [comment]
Idea: instead of [naively creating a web-socket](https://github.com/flutter/flutter/blob/2de61a0853f58b53673045149778bd87bc627039/packages/flutter_tools/lib/src/vmservice.dart#L29), we could have [Device.startApp](https://github.com/flutter/flutter/blob/7b2367ed5fa6cc596c2ae04d09364c4deed9d227/packages/flutter_tools/lib/src/run_hot.dart#L200) return a `StreamChannel` implementation that is capable of watching the status of `adb` and restarting it when necessary.

**Comment by eseidelGoogle** [comment]
Any update here?  Should we keep this in M2?

**Comment by johnmccutchan** [comment]
We are still waiting to get an updated trace from @Hixie. No one else has reported this issue. It is looking like it might be some kind of adb transport / port forwarding bug.

**Comment by Hixie** [comment]
@xster said the other day that he's seen this too, IIRC.

**Comment by Hixie** [comment]
(Sorry I haven't been using "flutter run" much in the last few days so haven't had a chance to catch this in action.)

**Comment by xster** [comment]
Can't really repro this but the few times I've seen it happen on a real device were all on iOS. The app itself closed as well while the 'session' was still running on the host which then quits. 

Presumably the app crashed somehow and there might be some modes of failure that doesn't output anything in the terminal.

**Comment by johnmccutchan** [comment]
@xster When the app closes / crashes we expect flutter tools to exit. This issue is tracking the case where the app is still running but flutter tools believe it exited.

**Comment by yjbanov** [comment]
BTW, here's a super-easy way to reproduce this:

* in one terminal window run `flutter run` and keep it running
* in _another_ terminal window run `adb kill-server`

**Comment by johnmccutchan** [comment]
@yjbanov that is really useful for attempting to work around the problem but it doesn't bring us any closer to the actual cause.

**Comment by yjbanov** [comment]
@johnmccutchan which comment are you referring to? My latest comment does not provide a workaround, but a way to reproduce the behavior. Granted, it does not prove that the cause of @Hixie's original problem was a flaky adb and so shielding our VM service connection from killed adb may not actually fix the original issue.

**Comment by Hixie** [comment]
Today I saw this but I know what caused it.

I was running `flutter run`, and in another window I ran `adb logcat`. That second command said:
```
adb server is out of date.  killing...
* daemon started successfully *
```
...and at that moment `flutter run` quit.

To clarify, I do not believe this explains other times I've seen it. I don't normally run `adb` directly.

**Comment by Hixie** [comment]
In fact in general running `adb logcat` in the background seems to trigger this even if you run `flutter run` second. Maybe I had an `adb logcat` process running somewhere in some window before?

**Comment by yjbanov** [comment]
This happened to me when I used two different versions of the Android SDK. When you run `adb` and it detects another version of `adb` running, it kills the other one, and that, as I mentioned in https://github.com/flutter/flutter/issues/8876#issuecomment-293658689, should lead to the incorrect "Application Stopped" behavior.

On top of that, we have a bug in `flutter_tools` (not sure if it's logged). We fail to detect the device when `flutter_tools` runs the `adb` command that prints `adb server is out of date.  killing...`. It thinks that `adb` in that line is the device ID. So the order in which you run `adb` and `flutter_tools` matters.

**Comment by Hixie** [comment]
Ah, yeah. Looks like I'm using the adb out of my engine repo but flutter is using the adb from the Android Studio install.

**Comment by Hixie** [comment]
I am going to close this for now. I'll open a new bug if I see this again and am sure it's not #9430.

**Comment by Hixie** [comment]
So I did just see this again and I definitely wasn't starting my own adb at the time.
Based on my inspection of the situation at the time, what happened is adb failed somehow. Certainly the process was no longer running.
If we're in verbose mode, can we run adb in some sort of verbose mode and when it dies (or when we lose our connection -- I did see "Lost connection to device"), output the most recent few hundred log lines from adb?
